crates/utils/src/interval_map/mod.rs

```diff
@@ -0,0 +1,1006 @@
+    fn update_max_bottom_up(&mut self, x: NodeIndex<Ix>) {
+        let mut p = x;
+        while !self.nref(p, Node::is_sentinel) {
+            self.nmut(p, Node::set_max(self.nref(p, Node::interval).high.clone()));
+            self.max_from(p, self.nref(p, Node::left));
+            self.max_from(p, self.nref(p, Node::right));
+            p = self.nref(p, Node::parent);
+        }
+    }
+
+    /// Updates a nodes value from a child node.
+    fn max_from(&mut self, x: NodeIndex<Ix>, c: NodeIndex<Ix>) {
+        if let Some(cmax) = self.nref(c, Node::sentinel).map(Node::max) {
+            let max = self.nref(x, Node::max).max(cmax).clone();
+            self.nmut(x, Node::set_max(max));
+        }
+    }
+
+    /// Finds the node with the minimum interval.
+    fn tree_minimum(&self, mut x: NodeIndex<Ix>) -> NodeIndex<Ix> {
+        while !self.left_ref(x, Node::is_sentinel) {
+            x = self.nref(x, Node::left);
+        }
+        x
+    }
+
+    /// Replaces one subtree as a child of its parent with another subtree.
+    fn transplant(&mut self, u: NodeIndex<Ix>, v: NodeIndex<Ix>) {
+        if self.parent_ref(u, Node::is_sentinel) {
+            self.root = v;
+        } else {
+            if self.is_left_child(u) {
+                self.parent_mut(u, Node::set_left(v));
+            } else {
+                self.parent_mut(u, Node::set_right(v));
+            }
+            self.update_max_bottom_up(self.nref(u, Node::parent));
+        }
+        self.nmut(v, Node::set_parent(self.nref(u, Node::parent)));
+    }
+
+    /// Checks if a node is a left child of its parent.
+    fn is_left_child(&self, node: NodeIndex<Ix>) -> bool {
+        self.parent_ref(node, Node::left) == node
+    }
+
+    /// Checks if a node is a right child of its parent.
+    fn is_right_child(&self, node: NodeIndex<Ix>) -> bool {
+        self.parent_ref(node, Node::right) == node
+    }
+
+    /// Updates nodes index after remove
+    fn update_idx(&mut self, old: NodeIndex<Ix>, new: NodeIndex<Ix>) {
+        if self.root == old {
+            self.root = new;
+        }
+        if self.nodes.get(new.index()).is_some() {
+            if !self.parent_ref(new, Node::is_sentinel) {
+                if self.parent_ref(new, Node::left) == old {
+                    self.parent_mut(new, Node::set_left(new));
+                } else {
+                    self.parent_mut(new, Node::set_right(new));
+                }
+            }
+            self.left_mut(new, Node::set_parent(new));
+            self.right_mut(new, Node::set_parent(new));
+        }
+    }
+}
+
+// Convenient methods for reference or mutate current/parent/left/right node
+#[allow(clippy::missing_docs_in_private_items)] // Trivial convenient methods
+#[allow(clippy::indexing_slicing)] // Won't panic since all the indices we used are inbound
+impl<'a, T, V, Ix> IntervalMap<T, V, Ix>
+where
+    Ix: IndexType,
+{
+    fn nref<F, R>(&'a self, node: NodeIndex<Ix>, op: F) -> R
```

Comment:
Does 'n' represent 'node'?

Reply:
yes

Reply:
I think using `node_ref` is more unified with other methods.

Reply:
done